File: sm3.cpp
#include <string.h>
#include "sm3.h"

static const uint32 T_J[] = {
    0x79cc4519, 0x7a879d8a, 0x7a879d8a, 0x7a879d8a
};

static const uint32 IV[] = {
    0x7380166f, 0x4914b2b9, 0x172442d7, 0xda8a0600,
    0xa96f30bc, 0x163138aa, 0xe38dee4d, 0xb0fb0e4e
};

#if WORD_SIZE == 8
static const word IPAD = 0x3636363636363636LL;
static const word OPAD = 0x5c5c5c5c5c5c5c5cLL;
#else
static const word IPAD = 0x36363636;
static const word OPAD = 0x5c5c5c5c;
#endif

inline uint32 rotl(uint32 x, int n) {
    return x << n | x >> (32 - n);
}

inline uint32 byte_int32(cbytes b) {
    return b[0] << 24 | b[1] << 16 | b[2] << 8 | b[3];
}

inline void int32_bytes(bytes out, uint32 x) {
    out[0] = (byte)(x >> 24);
    out[1] = (byte)(x >> 16);
    out[2] = (byte)(x >> 8);
    out[3] = (byte)x;
}

inline void int64_bytes(bytes out, uint64 x) {
    out[0] = (byte)(x >> 56);
    out[1] = (byte)(x >> 48);
    out[2] = (byte)(x >> 40);
    out[3] = (byte)(x >> 32);
    out[4] = (byte)(x >> 24);
    out[5] = (byte)(x >> 16);
    out[6] = (byte)(x >> 8);
    out[7] = (byte)x;
}

inline uint32 ff_j(uint32 x, uint32 y, uint32 z, uint32 j) {
    return (j >= 16)
           ? (x & y) | (x & z) | (y & z)
           : x ^ y ^ z;
}

inline uint32 gg_j(uint32 x, uint32 y, uint32 z, uint32 j) {
    return (j >= 16)
           ? (x & y) | (~x & z)
           : x ^ y ^ z;
}

inline uint32 p_0(uint32 x) {
    return x ^ rotl(x, 9) ^ rotl(x, 17);
}

inline uint32 p_1(uint32 x) {
    return x ^ rotl(x, 15) ^ rotl(x, 23);
}

inline void bytes_4int(uint32 *w, cbytes b) {
    w[0] = byte_int32(b);
    w[1] = byte_int32(b + 4);
    w[2] = byte_int32(b + 8);
    w[3] = byte_int32(b + 12);
}

static void get_w(uint32 *w, uint32 *w1, cbytes bi) {
    bytes_4int(w, bi);
    bytes_4int(w + 4, bi + 16);
    bytes_4int(w + 8, bi + 32);
    bytes_4int(w + 12, bi + 48);
    for (int j = 16; j < 68; j++) {
        w[j] = p_1(w[j - 16] ^ w[j - 9] ^ rotl(w[j - 3], 15))
               ^ rotl(w[j - 13], 7) ^ w[j - 6];
    }
    for (int j = 0; j < 64; j += 4) {
        w1[j] = w[j] ^ w[j + 4];
        w1[j + 1] = w[j + 1] ^ w[j + 5];
        w1[j + 2] = w[j + 2] ^ w[j + 6];
        w1[j + 3] = w[j + 3] ^ w[j + 7];
    }
}

static uint32 *cf(uint32 *v_out, const uint32 *vi, cbytes bi) {
    uint32 w[68], w1[64];
    get_w(w, w1, bi);
    uint32 a = vi[0];
    uint32 b = vi[1];
    uint32 c = vi[2];
    uint32 d = vi[3];
    uint32 e = vi[4];
    uint32 f = vi[5];
    uint32 g = vi[6];
    uint32 h = vi[7];

    for (int j = 0; j < 64; j++) {
        uint32 s1 = rotl(rotl(a, 12) + e + rotl(T_J[j >> 4], (j & 31)), 7);
        uint32 s2 = s1 ^ rotl(a, 12);
        uint32 t1 = ff_j(a, b, c, j) + d + s2 + w1[j];
        uint32 t2 = gg_j(e, f, g, j) + h + s1 + w[j];
        d = c;
        c = rotl(b, 9);
        b = a;
        a = t1;
        h = g;
        g = rotl(f, 19);
        f = e;
        e = p_0(t2);
    }

    v_out[0] = a ^ vi[0];
    v_out[1] = b ^ vi[1];
    v_out[2] = c ^ vi[2];
    v_out[3] = d ^ vi[3];
    v_out[4] = e ^ vi[4];
    v_out[5] = f ^ vi[5];
    v_out[6] = g ^ vi[6];
    v_out[7] = h ^ vi[7];
    return v_out;
}

static int padding(bytes dst, cbytes src, int srclen, size_t total) {
    memcpy(dst, src, srclen);
    dst[srclen++] = '\x80';
    int lenpos = (srclen <= 56) ? 56 : 120;
    int64_bytes(dst + lenpos, (uint64)total << 3);
    return lenpos + 8;
}

static void sm3_update(uint32 *state, cbytes src, size_t srclen, const uint32 *iv) {
    if (iv == NULL)
        iv = IV;
    for (size_t i = 0; i < srclen; i += 64) {
        iv = cf(state, iv, src + i);
    }
}

static void sm3_finish(bytes hash, cbytes src, size_t srclen, size_t total, const uint32 *iv) {
    uint32 state[8];
    size_t len64 = srclen & (SIZE_MAX - 63);
    for (size_t i = 0; i < len64; i += 64) {
        iv = cf(state, iv, src + i);
    }
    byte buff[128] = {0};
    int lastlen = padding(buff, src + len64, (int)(srclen - len64), total);
    for (int i = 0; i < lastlen; i += 64) {
        iv = cf(state, iv, buff + i);
    }
    int32_bytes(hash, state[0]);
    int32_bytes(hash + 4, state[1]);
    int32_bytes(hash + 8, state[2]);
    int32_bytes(hash + 12, state[3]);
    int32_bytes(hash + 16, state[4]);
    int32_bytes(hash + 20, state[5]);
    int32_bytes(hash + 24, state[6]);
    int32_bytes(hash + 28, state[7]);
}

bytes sm3_hash(bytes hash, cbytes src, size_t srclen) {
    sm3_finish(hash, src, srclen, srclen, IV);
    return hash;
}

#define BLOCK_SIZE 64
#define DIGEST_SIZE 32
static void xorpad(word *out, const word *src, const word pad) {
    for (int i = 0; i < BLOCK_SIZE / WORD_SIZE; i += 4) {
        out[i] = src[i] ^ pad;
        out[i + 1] = src[i + 1] ^ pad;
        out[i + 2] = src[i + 2] ^ pad;
        out[i + 3] = src[i + 3] ^ pad;
    }
}

bytes sm3_hmac(bytes hmac, cbytes key, size_t keylen, cbytes src, size_t srclen) {
    byte keyblk[BLOCK_SIZE] = {0};
    if (keylen > BLOCK_SIZE) {
        sm3_hash(keyblk, src, srclen);
    } else {
        memcpy(keyblk, key, keylen);
    }

    byte inner[BLOCK_SIZE];
    byte outer[BLOCK_SIZE + DIGEST_SIZE];
    xorpad((word *)inner, (word *)keyblk, IPAD);
    xorpad((word *)outer, (word *)keyblk, OPAD);

    uint32 state[8];
    sm3_update(state, inner, BLOCK_SIZE, NULL);
    sm3_finish(outer + BLOCK_SIZE, src, srclen, BLOCK_SIZE + srclen, state);
    sm3_hash(hmac, outer, BLOCK_SIZE + DIGEST_SIZE);
    return hmac;
}

File: sm3.py
''' SM3 Hash '''
from struct import pack, unpack, unpack_from
try:
    from _sm3 import sm3_hash, sm3_hmac
    FAST = True
except ImportError:
    FAST = False

IV = (
    0x7380166f, 0x4914b2b9, 0x172442d7, 0xda8a0600,
    0xa96f30bc, 0x163138aa, 0xe38dee4d, 0xb0fb0e4e
)

T_J =  (0x79cc4519, 0x7a879d8a, 0x7a879d8a, 0x7a879d8a)


def rotl(value, cnt):
    ''' 32bits left shift'''
    return value << cnt & 0xFFFFFFFF | value >> (32 - cnt)

def xor256(data1, data2):
    ''' 256 bits xor '''
    idata1 = unpack('4Q', data1)
    idata2 = unpack('4Q', data2)
    return pack('4Q',
        idata1[0] ^ idata2[0],
        idata1[1] ^ idata2[1],
        idata1[2] ^ idata2[2],
        idata1[3] ^ idata2[3]
    )

def _xorpad(data, pad):
    ''' Xor bytes of data and pad '''
    idata = unpack('8Q', data)
    return pack('8Q',
        idata[0] ^ pad, idata[1] ^ pad,
        idata[2] ^ pad, idata[3] ^ pad,
        idata[4] ^ pad, idata[5] ^ pad,
        idata[6] ^ pad, idata[7] ^ pad
    )

def _get_w(data, offset):
    w_array = list(unpack_from('>16I', data, offset))
    for j in range(16, 68):
        xxx = w_array[j-16] ^ w_array[j-9] ^ rotl(w_array[j-3], 15)
        w_array.append(xxx ^ rotl(xxx, 15) ^ rotl(xxx, 23)  # p1(xxx)
                       ^ rotl(w_array[j-13], 7) ^ w_array[j-6])
    w1_array = [w_array[j] ^ w_array[j+4] for j in range(64)]
    return w_array, w1_array


def _cf(vector, data, offset=0):
    # pylint: disable=too-many-locals
    w_array, w1_array = _get_w(data, offset)
    aaa, bbb, ccc, ddd, eee, fff, ggg, hhh = vector

    for j in range(0, 16):
        ss1 = rotl(((rotl(aaa, 12)) + eee + (rotl(T_J[j>>4], j & 31))) & 0xFFFFFFFF, 7)
        ss2 = ss1 ^ rotl(aaa, 12)
        tt1 = ((aaa ^ bbb ^ ccc) + ddd + ss2 + w1_array[j]) & 0xFFFFFFFF
        tt2 = ((eee ^ fff ^ ggg) + hhh + ss1 + w_array[j]) & 0xFFFFFFFF
        ddd = ccc
        ccc = rotl(bbb, 9)
        bbb = aaa
        aaa = tt1
        hhh = ggg
        ggg = rotl(fff, 19)
        fff = eee
        eee = tt2 ^ rotl(tt2, 9) ^ rotl(tt2, 17)  # p0(t2)

    for j in range(16, 64):
        ss1 = rotl(((rotl(aaa, 12)) + eee + (rotl(T_J[j>>4], j & 31))) & 0xFFFFFFFF, 7)
        ss2 = ss1 ^ rotl(aaa, 12)
        tt1 = ((aaa & bbb | aaa & ccc | bbb & ccc) + ddd + ss2 + w1_array[j]) & 0xFFFFFFFF
        tt2 = ((eee & fff | ~eee & ggg) + hhh + ss1 + w_array[j]) & 0xFFFFFFFF
        ddd = ccc
        ccc = rotl(bbb, 9)
        bbb = aaa
        aaa = tt1
        hhh = ggg
        ggg = rotl(fff, 19)
        fff = eee
        eee = tt2 ^ rotl(tt2, 9) ^ rotl(tt2, 17)  # p0(t2)

    return (
        aaa ^ vector[0],
        bbb ^ vector[1],
        ccc ^ vector[2],
        ddd ^ vector[3],
        eee ^ vector[4],
        fff ^ vector[5],
        ggg ^ vector[6],
        hhh ^ vector[7]
    )


def _padding(data, total):
    padlen = 56 - ((total + 1) & 63)
    if padlen < 0:
        padlen += 64
    data += pack(f'>B{padlen}xQ', 128, total<<3)
    return data


def _sm3_hash(data):
    ''' SM3 Hash '''
    vector = IV
    datalen = len(data)
    len64 = datalen - (datalen & 63)
    for pos in range(0, len64, 64):
        vector = _cf(vector, data, pos)
    data = _padding(data[len64:], datalen)
    for pos in range(0, len(data), 64):
        vector = _cf(vector, data, pos)
    return pack('>8I', *vector)


_BLOCK_SIZE = 64
_IPAD = 0x3636363636363636
_OPAD = 0x5C5C5C5C5C5C5C5C
def _sm3_hmac(key, data):
    ''' SM3 HMac Hash '''
    if len(key) > _BLOCK_SIZE:
        key = sm3_hash(key)
    key = key.ljust(_BLOCK_SIZE, b'\0')
    hashv = sm3_hash(_xorpad(key, _IPAD) + data)
    return sm3_hash(_xorpad(key, _OPAD) + hashv)


def pbkdf2(password, salt, iterations, dklen=32):
    ''' Password-based key derivation function 2. '''
    result = bytearray()
    blk_cnt = (dklen + 31) >> 5
    for idx in range(1, blk_cnt + 1):
        salt_i4 = salt + pack('>I', idx)
        hmac_result = sm3_hmac(password, salt_i4)
        xor_result = hmac_result
        for _ in range(1, iterations):
            hmac_result = sm3_hmac(password, hmac_result)
            xor_result = xor256(xor_result, hmac_result)
        # Concatenate the blocks
        result.extend(xor_result)
    return result[:dklen]


class SM3:
    ''' SM3 Hash Class '''
    __slots__ = ['buffer', 'length', 'outbuf']

    def __init__(self, data=b''):
        self.buffer = bytearray(data)
        self.length = len(data)
        self.outbuf = IV
        len64 = len(self.buffer)
        len64 = len64 - (len64 & 63)
        for pos in range(0, len64, 64):
            self.outbuf = _cf(self.outbuf, self.buffer, pos)
        self.buffer = self.buffer[len64:]

    @classmethod
    def new(cls, data=b''):
        ''' Create a new hashing object. '''
        return cls(data)

    @property
    def name(self):
        ''' The hash algorithm being used by this object. '''
        return 'sm3'

    @property
    def digest_size(self):
        ''' The size of the resulting hash in bytes.. '''
        return 32

    @property
    def block_size(self):
        ''' The block size of the hash algorithm in bytes. '''
        return 64

    def copy(self):
        ''' Return a copy of the current hash object. '''
        obj = self.__class__()
        obj.buffer = self.buffer.copy()
        obj.length = self.length
        obj.outbuf = self.outbuf
        return obj

    def update(self, data):
        ''' Updat the current digest with an additional string. '''
        self.buffer.extend(data)
        self.length += len(data)
        len64 = len(self.buffer)
        len64 = len64 - (len64 & 63)
        for pos in range(0, len64, 64):
            self.outbuf = _cf(self.outbuf, self.buffer, pos)
        self.buffer = self.buffer[len64:]

    def digest(self):
        ''' Return the current digest value as bytes. '''
        self.buffer = _padding(self.buffer, self.length)
        for pos in range(0, len(self.buffer), 64):
            self.outbuf = _cf(self.outbuf, self.buffer, pos)
        self.outbuf = bytearray()
        return pack('>8I', *self.outbuf)

    def hexdigest(self):
        ''' Return the current digest as hexadecimal string. '''
        return self.digest().hex()


if not FAST:
    sm3_hash = _sm3_hash
    sm3_hmac = _sm3_hmac
